File: sqlens-backend/app/schemas/query_history.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class QueryHistoryItemResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: str = Field(min_length=1)
    query: str = Field(min_length=1)
    status: Literal["success", "error"]
    executed_at: datetime = Field(serialization_alias="executedAt")
    execution_time_ms: float | None = Field(default=None, ge=0, serialization_alias="executionTimeMs")
    row_count: int | None = Field(default=None, ge=0, serialization_alias="rowCount")


class QueryHistoryListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[QueryHistoryItemResponse]

File: sqlens-frontend/lib/api/contracts/query-history.ts
import { z } from "zod";

export const queryHistoryStatusSchema = z.enum(["success", "error"]);

/**
 * Transport contract for one persisted execution owned by the authenticated
 * user. Optional metrics are omitted when the SQL engine cannot provide them.
 */
export const queryHistoryItemSchema = z
  .object({
    id: z.string().min(1),
    query: z.string().min(1),
    status: queryHistoryStatusSchema,
    executedAt: z.string().datetime({ offset: true }),
    executionTimeMs: z.number().finite().nonnegative().optional(),
    rowCount: z.number().int().nonnegative().optional(),
  })
  .strict();

export const queryHistoryListResponseSchema = z
  .object({
    items: z.array(queryHistoryItemSchema),
  })
  .strict();

export type QueryHistoryItem = z.infer<typeof queryHistoryItemSchema>;
export type QueryHistoryListResponse = z.infer<
  typeof queryHistoryListResponseSchema
>;
